### ISSUE TITLE
"Human names" for terminals

### ISSUE THREAD

**Issue by remexre** [melt-umn/silver#205]
It might be nice for error messages to use "human-readable" names for terminals; these could probably be autogenerated in the case that the terminal is a constant string, but would be needed if the terminal is a regex.

For the grammar:

```silver
terminal Float_Literal_t /[\-]?[0-9]+([\.][0-9]*)?/ lexer classes {LITERAL} named "a number";
terminal Plus_t             '+'  association = left lexer classes {OPERATOR}; -- name inferred to be "+"

nonterminal Expr named "an expression";

concrete production lit
top::Expr ::= l::Float_Literal_t
{ ... }

concrete production plus
top::Expr ::= l::Expr '+' r::Float_Literal_t
{ ... }
```

parsing:

`1 + + 2`

would (eventually) give something like

```
error: expected a number, found `+`
 --> example.pluslang:1:5
  |
1 |     1 + + 2
  |         ^
```

rather than the current (much harder-to-read) errors

**Comment by krame505** [comment]
Seems like a nice idea.  Although, it would have to be `expected <a long list of things>, found '+'` unless we added some sort of prediction mechanism to Copper.

**Comment by remexre** [comment]
Maybe also have some sort of concept of "alias groups," so adding:

```silver
alias group "a binary operator" {
  '+', '-', '*', '/', '!', '&&', '||'
};
alias group "an unary operator" {
  '+', '-', '!'
};
alias group "a literal" {
  Float_Literal_t, String_Literal_t
};
```

and then we could transform something like

```
expected one of ['+', '-', '!', Float_Literal_t, String_Literal_t, Other_Thing_t]
```

into

```
expected an unary operator, a literal, or one of [Other_Thing_t]
```

I _think_ that doing this optimally (fewest number of things in the "expected" list) might be NP-hard (equiv. to knapsack problem), but I haven't gotten to that part of 4011 yet, so it might not be. And if we do it greedily, that's probably "good enough".

**Comment by krame505** [comment]
I wouldn't be too concerned about performance or how we select groups - these would be mostly non-overlapping, anyway.

**Comment by ericvanwyk** [comment]
@schwerdf does Copper already have some support for this?  I vaguely recall some discussion of specifying more intuitive names for terminals so that Copper could report these on syntax errors instead of the fully qualified name of the terminals.

**Comment by schwerdf** [comment]
Yes. Copper terminals, and other grammar objects, have a `displayName` field (corresponding to the `<PP>` elements in the XML schema) to hold the "pretty print" name of the object. I thought Silver already used this to render the fully-qualified names of grammar symbols with colons instead of underscores.

As for custom error messages in general, you could implement a custom Copper logger on the Silver side that would render the messages however you need, including any NP-hard set computation.

**Comment by tedinski** [comment]
We had planned to use that new copper feature to automatically make keywords (`terminal 'foo'`) get named `'foo'` instead of `grammar:name:Foo_t` but the structure of the Silver compiler made this slightly more difficult than I wished it was, and I never went back and did it.

RE: "alias groups" we ought to be able to do something like this with lexer classes.

**Comment by remexre** [comment]
So I have a really basic first pass in [`a1ce700`](https://github.com/melt-umn/silver/commit/a1ce700e1ce6d44dc41c294a87a54920e331b559) (on the branch [`feature/better-errors`](https://github.com/melt-umn/silver/compare/feature/better-errors)):

![](https://i.imgur.com/jWbZUfP.png)

**Comment by ericvanwyk** [comment]
Looks like progress!

**Comment by tedinski** [comment]
Was this solved by #221, should this issue be closed @remexre ?

**Comment by remexre** [comment]
Oh, yep.
